code/modules/GOAP/GOAPPlanner.dm

```diff
@@ -0,0 +1,134 @@
+
+/datum/goap_planner
+	var/list/stored_nodes = list()
+
+/datum/goap_planner/proc/Plan(atom/agent, list/actions, list/worldstate, list/goal)
+	for(var/a in actions)
+		var/datum/goap_action/GA = a
+		GA.Reset()
+		CHECK_TICK
+
+	var/list/usable_actions = list()
+
+	for(var/a in actions)
+		var/datum/goap_action/GA = a
+		goap_debug("CHECKING ACTION [GA]")
+		if(GA.cooldown)
+			if(GA.OnCooldown(agent, worldstate))
+				continue
+		if(GA.AdvancedPreconditions(agent, worldstate))
+			usable_actions += GA
+			goap_debug("USABLE ACTION [GA]")
+		CHECK_TICK
+
+	//Oh god, trees! I hate this!
+
+	var/list/plan_tree = list()
+
+	var/datum/goap_plan_node/start = new()
+	start.state = worldstate
+
+	plan_tree = BuildPossiblePlans(start, usable_actions, goal)
+
+	if(!plan_tree.len)
+		goap_debug("NO PLAN CREATED")
+		return null
+
+	//Cheapest path
+	var/datum/goap_plan_node/cheapest
+	for(var/node in plan_tree)
+		var/datum/goap_plan_node/N = node
+		goap_debug("NODE IS [N.action] COST IS [N.cost]")
+		if(!cheapest || N.cost < cheapest.cost)
+			goap_debug("CURRENT CHEAPEST [N.cost]")
+			cheapest = N
+		CHECK_TICK
+
+	//Go up the tree, from the cheapest bottom leaf
+	var/list/plan = list()
+	var/datum/goap_plan_node/climber = cheapest
+	while(climber)
+		if(climber.action)
+			plan += climber.action
+		climber = climber.parent
+		CHECK_TICK
+	return plan
+
+
+//Builds the tree of actions
+//Goes through them all.
+//ALL OF THEM, we have to in order to find the cheapest path
+//TODO: remove recursion it's icky and slow
```

Comment:
@RemieRichards should we care or not because i don't feel like thinking up a solution rn